From ff0cd3fb90b5ffb08aa233d2ba4bc5be63ed0843 Mon Sep 17 00:00:00 2001
From: Chau Ly <chaul@amperecomputing.com>
Date: Wed, 19 Jul 2023 08:42:18 +0000
Subject: [PATCH] chassis: Methods to PhysicalSecurity's properties

This commit handles GET method to IntrusionSensorReArm property under
PhysicalSecurity. Moreover, Redfish Data Model Specification says about
IntrusionSensor that "If the IntrusionSensorReArm property contains
Manual a client may set this property to Normal to reset the physical
security state". This will handle PATCH to IntrusionSensor under
PhysicalSecurity.

This commit also deprecates IntrusionSensorNumber property as Redfish
schema v1.22 specified "IntrusionSensorNumber has been deprecated in
order to allow for multiple physical sensors to construct this object".
Moreover, it was previously hardcoded to 1, so there are no clients
that could be using this, and per COMPATIBILITY.md, this is not a
compatibility break.

Redfish.md is updated to add PhysicalSecurity to under
`/redfish/v1/Chassis/{ChassisId}/`.

Tested: On Ampere Mt.Mitchell using dbus-sensors intrusionsensor with
Aspeed2600_Hwmon method, rearm mode is Manual. `chassis` here is the
chassis associated with the chassis intrusion sensor on DBus by the
monitored_by/monitoring association.

- Normal situation:

$ curl -X GET -u root:0penBmc --insecure \
     https://${BMC_IP}/redfish/v1/Chassis/chassis

...
"PhysicalSecurity": {
    "IntrusionSensor": "Normal",
    "IntrusionSensorReArm": "Manual"
},
...

- Chassis intrusion detected, the above result will be:

"PhysicalSecurity": {
    "IntrusionSensor": "HardwareIntrusion",
    "IntrusionSensorReArm": "Manual"
}

- Manually rearm the sensor:

$ curl -X PATCH --user root:0penBmc -H \
"Content-Type: application/json" --insecure \
https://${BMC_IP}/redfish/v1/Chassis/chassis -d \
'{ "PhysicalSecurity": { "IntrusionSensor": "Normal"} }'

{
  "@Message.ExtendedInfo": [
    {
      "@odata.type": "#Message.v1_1_1.Message",
      "Message": "The request completed successfully.",
      "MessageArgs": [],
      "MessageId": "Base.1.16.0.Success",
      "MessageSeverity": "OK",
      "Resolution": "None"
    }
  ]
}

The result will be as in normal situation

- If rearm the sensor from an other chassis

$ curl -X PATCH --user root:0penBmc -H \
"Content-Type: application/json" --insecure \
https://${BMC_IP}/redfish/v1/Chassis/<another-chassis> -d \
'{ "PhysicalSecurity": { "IntrusionSensor": "Normal"} }'

It returns Internal error

Upstream-Status: Pending

Change-Id: If50332a1e55f2bafcf395e0d313b7d58b35c55cd
Signed-off-by: Chau Ly <chaul@amperecomputing.com>
---
 redfish-core/lib/chassis.hpp | 303 ++++++++++++++++++++++++++++-------
 1 file changed, 248 insertions(+), 55 deletions(-)

diff --git a/redfish-core/lib/chassis.hpp b/redfish-core/lib/chassis.hpp
index aab95bd2..9e7da207 100644
--- a/redfish-core/lib/chassis.hpp
+++ b/redfish-core/lib/chassis.hpp
@@ -48,6 +48,58 @@
 
 namespace redfish
 {
+/**
+ * @brief Convert chassis intrusion sensor status string on DBus to Redfish
+ *
+ * @param[in] status - Status property string on DBus.
+ *
+ * @return An optional of chassis::IntrusionSensor enum.
+ */
+inline std::optional<chassis::IntrusionSensor> dbusChassisIntrusionStatusToRf(
+    std::string_view status)
+{
+    if (status == "xyz.openbmc_project.Chassis.Intrusion.Status.Normal")
+    {
+        return chassis::IntrusionSensor::Normal;
+    }
+    if (status ==
+        "xyz.openbmc_project.Chassis.Intrusion.Status.HardwareIntrusion")
+    {
+        return chassis::IntrusionSensor::HardwareIntrusion;
+    }
+    if (status ==
+        "xyz.openbmc_project.Chassis.Intrusion.Status.TamperingDetected")
+    {
+        return chassis::IntrusionSensor::TamperingDetected;
+    }
+    if (status == "xyz.openbmc_project.Chassis.Intrusion.Status.Unknown")
+    {
+        return std::nullopt;
+    }
+    return chassis::IntrusionSensor::Invalid;
+}
+
+/**
+ * @brief Convert chassis intrusion sensor rearm mode string on DBus to Redfish
+ *
+ * @param[in] rearmMode - Rearm property string on DBus.
+ *
+ * @return A value of chassis::IntrusionSensorReArm enum.
+ */
+inline chassis::IntrusionSensorReArm dbusChassisIntrusionRearmModeToRf(
+    std::string_view rearmMode)
+{
+    if (rearmMode == "xyz.openbmc_project.Chassis.Intrusion.RearmMode.Manual")
+    {
+        return chassis::IntrusionSensorReArm::Manual;
+    }
+    if (rearmMode ==
+        "xyz.openbmc_project.Chassis.Intrusion.RearmMode.Automatic")
+    {
+        return chassis::IntrusionSensorReArm::Automatic;
+    }
+    return chassis::IntrusionSensorReArm::Invalid;
+}
 
 inline chassis::ChassisType translateChassisTypeToRedfish(
     const std::string_view& chassisType)
@@ -195,51 +247,144 @@ inline void getChassisState(std::shared_ptr<bmcweb::AsyncResp> asyncResp)
 }
 
 /**
- * Retrieves physical security properties over dbus
+ * Translate DBus property values to Redfish and fill PhysicalSecurity's data
+ */
+inline void handlePhysicalSecurityProperties(
+    crow::Response& resp, const dbus::utility::DBusPropertiesMap& propertiesMap)
+{
+    std::string status;
+    std::string rearmMode;
+
+    const bool success = sdbusplus::unpackPropertiesNoThrow(
+        dbus_utils::UnpackErrorPrinter(), propertiesMap, "Status", status,
+        "Rearm", rearmMode);
+
+    if (!success)
+    {
+        messages::internalError(resp);
+        return;
+    }
+
+    std::optional<chassis::IntrusionSensor> intrusionSensor =
+        dbusChassisIntrusionStatusToRf(status);
+
+    if (!intrusionSensor)
+    {
+        BMCWEB_LOG_WARNING("Unknown ChassisIntrusionStatus: {}", status);
+    }
+    else
+    {
+        if (*intrusionSensor == chassis::IntrusionSensor::Invalid)
+        {
+            BMCWEB_LOG_ERROR("Invalid ChassisIntrusionStatus: {}", status);
+            messages::internalError(resp);
+            return;
+        }
+        resp.jsonValue["PhysicalSecurity"]["IntrusionSensor"] =
+            *intrusionSensor;
+        resp.jsonValue["PhysicalSecurity"]
+                      ["IntrusionSensor@Redfish.AllowableValues"] =
+            nlohmann::json::array_t({"Normal"});
+    }
+
+    chassis::IntrusionSensorReArm intrusionSensorReArm =
+        dbusChassisIntrusionRearmModeToRf(rearmMode);
+
+    if (intrusionSensorReArm == chassis::IntrusionSensorReArm::Invalid)
+    {
+        BMCWEB_LOG_ERROR("Invalid ChassisIntrusionRearmMode: {}", rearmMode);
+        messages::internalError(resp);
+        return;
+    }
+    resp.jsonValue["PhysicalSecurity"]["IntrusionSensorReArm"] =
+        intrusionSensorReArm;
+}
+
+/**
+ * Handle Set/Get properties call to intrusion sensor object
  */
-inline void handlePhysicalSecurityGetSubTree(
+inline void callToIntrusionSensor(
     const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+    const dbus::utility::MapperEndPoints& monitoringEPs, bool callSet,
     const boost::system::error_code& ec,
     const dbus::utility::MapperGetSubTreeResponse& subtree)
 {
-    if (ec)
+    if (ec || subtree.empty())
     {
-        // do not add err msg in redfish response, because this is not
-        //     mandatory property
-        BMCWEB_LOG_INFO("DBUS error: no matched iface {}", ec);
+        if (callSet)
+        {
+            BMCWEB_LOG_ERROR("DBUS error: no matched iface");
+            messages::internalError(asyncResp->res);
+        }
+        else
+        {
+            // this is not a mandatory property
+            BMCWEB_LOG_INFO("DBUS error: no matched iface");
+        }
         return;
     }
-    // Iterate over all retrieved ObjectPaths.
-    for (const auto& object : subtree)
+
+    // There should be only one place implementing this interface
+    if (subtree.size() > 1)
     {
-        if (!object.second.empty())
+        BMCWEB_LOG_ERROR("Found more than one Intrusion Sensor DBus objects");
+        messages::internalError(asyncResp->res);
+        return;
+    }
+
+    const auto& [objPath, serviceMap] = subtree[0];
+
+    auto found = std::find(monitoringEPs.begin(), monitoringEPs.end(), objPath);
+    if (found == monitoringEPs.end())
+    {
+        if (callSet)
         {
-            const auto& service = object.second.front();
+            BMCWEB_LOG_ERROR(
+                "No intrusion sensor associated with this chassis");
+            messages::internalError(asyncResp->res);
+        }
+        return;
+    }
 
-            BMCWEB_LOG_DEBUG("Get intrusion status by service ");
+    if (serviceMap.empty())
+    {
+        BMCWEB_LOG_ERROR(
+            "Error getting service map of Intrusion Sensor object");
+        messages::internalError(asyncResp->res);
+        return;
+    }
 
-            dbus::utility::getProperty<std::string>(
-                service.first, object.first,
-                "xyz.openbmc_project.Chassis.Intrusion", "Status",
-                [asyncResp](const boost::system::error_code& ec1,
-                            const std::string& value) {
-                    if (ec1)
-                    {
-                        // do not add err msg in redfish response, because this
-                        // is not
-                        //     mandatory property
-                        BMCWEB_LOG_ERROR("DBUS response error {}", ec1);
-                        return;
-                    }
-                    asyncResp->res.jsonValue["PhysicalSecurity"]
-                                            ["IntrusionSensorNumber"] = 1;
-                    asyncResp->res
-                        .jsonValue["PhysicalSecurity"]["IntrusionSensor"] =
-                        value;
-                });
+    const auto& service = serviceMap.front().first;
 
-            return;
-        }
+    const std::string intrusionInterface =
+        "xyz.openbmc_project.Chassis.Intrusion";
+
+    if (callSet)
+    {
+        const std::string normalStatusDbusStr =
+            "xyz.openbmc_project.Chassis.Intrusion.Status.Normal";
+
+        sdbusplus::asio::setProperty(
+            *crow::connections::systemBus, service, objPath, intrusionInterface,
+            "Status", normalStatusDbusStr,
+            [asyncResp](const boost::system::error_code& ec1) {
+                if (ec1)
+                {
+                    BMCWEB_LOG_ERROR("DBUS error: failed to set property");
+                    messages::internalError(asyncResp->res);
+                    return;
+                }
+                messages::success(asyncResp->res);
+            });
+    }
+    else
+    {
+        sdbusplus::asio::getAllProperties(
+            *crow::connections::systemBus, service, objPath, intrusionInterface,
+            [asyncResp](const boost::system::error_code&,
+                        const dbus::utility::DBusPropertiesMap& propertiesMap) {
+                handlePhysicalSecurityProperties(asyncResp->res, propertiesMap);
+            });
     }
 }
 
@@ -539,6 +684,37 @@ inline void handleChassisProperties(
     }
 }
 
+/**
+ * Handle endpoints from getting the monitoring objects of the chassis
+ */
+inline void handleChassisMonitoringEndpoints(
+    const std::shared_ptr<bmcweb::AsyncResp>& asyncResp, bool callSet,
+    const boost::system::error_code& ec,
+    const dbus::utility::MapperEndPoints& endpoints)
+{
+    if (ec || endpoints.empty())
+    {
+        if (callSet)
+        {
+            BMCWEB_LOG_ERROR("DBUS error: no endpoint found");
+            messages::internalError(asyncResp->res);
+        }
+        else
+        {
+            BMCWEB_LOG_INFO("DBUS error: no endpoint found");
+        }
+        return;
+    }
+    constexpr std::array<std::string_view, 1> intrusionInterfaces = {
+        "xyz.openbmc_project.Chassis.Intrusion"};
+
+    sdbusplus::message::object_path path("/xyz/openbmc_project");
+
+    dbus::utility::getSubTree(
+        path, 0, intrusionInterfaces,
+        std::bind_front(callToIntrusionSensor, asyncResp, endpoints, callSet));
+}
+
 inline void handleChassisGetSubTree(
     const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
     const std::string& chassisId, const boost::system::error_code& ec,
@@ -709,6 +885,14 @@ inline void handleChassisGetSubTree(
             }
         }
 
+        sdbusplus::message::object_path chassisMonitoredPath =
+            path + "/monitored_by";
+
+        // Get PhysicalSecurity
+        dbus::utility::getAssociationEndPoints(
+            chassisMonitoredPath,
+            std::bind_front(handleChassisMonitoringEndpoints, asyncResp,
+                            false));
         return;
     }
 
@@ -729,19 +913,12 @@ inline void handleChassisGet(
     dbus::utility::getSubTree(
         "/xyz/openbmc_project/inventory", 0, chassisInterfaces,
         std::bind_front(handleChassisGetSubTree, asyncResp, chassisId));
-
-    constexpr std::array<std::string_view, 1> interfaces2 = {
-        "xyz.openbmc_project.Chassis.Intrusion"};
-
-    dbus::utility::getSubTree(
-        "/xyz/openbmc_project", 0, interfaces2,
-        std::bind_front(handlePhysicalSecurityGetSubTree, asyncResp));
 }
 
 inline void handleChassisPatch(
     App& app, const crow::Request& req,
     const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
-    const std::string& param)
+    const std::string& chassisId)
 {
     if (!redfish::setUpRedfishRoute(app, req, asyncResp))
     {
@@ -749,26 +926,23 @@ inline void handleChassisPatch(
     }
     std::optional<bool> locationIndicatorActive;
     std::optional<std::string> indicatorLed;
+    std::optional<std::string> intrusionSensor;
 
-    if (param.empty())
+    if (chassisId.empty())
     {
         return;
     }
 
-    if (!json_util::readJsonPatch(                             //
-            req, asyncResp->res,                               //
-            "IndicatorLED", indicatorLed,                      //
-            "LocationIndicatorActive", locationIndicatorActive //
+    if (!json_util::readJsonPatch(                              //
+            req, asyncResp->res,                                //
+            "IndicatorLED", indicatorLed,                       //
+            "LocationIndicatorActive", locationIndicatorActive, //
+            "PhysicalSecurity/IntrusionSensor", intrusionSensor //
             ))
     {
         return;
     }
 
-    // TODO (Gunnar): Remove IndicatorLED after enough time has passed
-    if (!locationIndicatorActive && !indicatorLed)
-    {
-        return; // delete this when we support more patch properties
-    }
     if (indicatorLed)
     {
         asyncResp->res.addHeader(
@@ -776,13 +950,12 @@ inline void handleChassisPatch(
             "299 - \"IndicatorLED is deprecated. Use LocationIndicatorActive instead.\"");
     }
 
-    const std::string& chassisId = param;
-
     dbus::utility::getSubTree(
         "/xyz/openbmc_project/inventory", 0, chassisInterfaces,
-        [asyncResp, chassisId, locationIndicatorActive,
-         indicatorLed](const boost::system::error_code& ec,
-                       const dbus::utility::MapperGetSubTreeResponse& subtree) {
+        [asyncResp, chassisId, locationIndicatorActive, indicatorLed,
+         intrusionSensor](
+            const boost::system::error_code& ec,
+            const dbus::utility::MapperGetSubTreeResponse& subtree) {
             if (ec)
             {
                 BMCWEB_LOG_ERROR("DBUS response error {}", ec);
@@ -855,6 +1028,26 @@ inline void handleChassisPatch(
                                                   "IndicatorLED");
                     }
                 }
+                if (intrusionSensor)
+                {
+                    if (*intrusionSensor != "Normal")
+                    {
+                        BMCWEB_LOG_ERROR(
+                            "IntrusionSensor property only accepts Normal to reset the physical security state");
+                        messages::propertyValueIncorrect(asyncResp->res,
+                                                         "IntrusionSensor",
+                                                         *intrusionSensor);
+                        return;
+                    }
+
+                    sdbusplus::message::object_path chassisMonitoredPath(
+                        path + "/monitored_by");
+
+                    dbus::utility::getAssociationEndPoints(
+                        chassisMonitoredPath,
+                        std::bind_front(handleChassisMonitoringEndpoints,
+                                        asyncResp, true));
+                }
                 return;
             }
 
